src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -263,129 +263,72 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
+      //
+      // NOTE: Unlike simple aggregate functions (e.g., `Max`, `Min`, etc.), `Count` must be
+      // replaced with a `Sum` to sum up the partial counts returned by TiKV.
+      //
+      // NOTE: All `Average`s should have already been rewritten into `Sum`s and `Count`s by the
+      // `TiAggregation` pattern extractor.
+
+      // An attribute referring to the partial aggregation results returned by TiKV.
+      val partialResultRef = aliasPushedPartialResult(aggExpr).toAttribute
+
       aggExpr.aggregateFunction match {
-        // here aggExpr is the original AggregationExpression
-        // and will be pushed down to TiKV
-        case Max(_)   => newAggregate(Max(toAlias(aggExpr).toAttribute), aggExpr)
-        case Min(_)   => newAggregate(Min(toAlias(aggExpr).toAttribute), aggExpr)
-        case Count(_) => newAggregate(Sum(toAlias(aggExpr).toAttribute), aggExpr)
-        case Sum(_)   => newAggregate(Sum(toAlias(aggExpr).toAttribute), aggExpr)
-        case First(_, ignoreNullsExpr) =>
-          newAggregate(First(toAlias(aggExpr).toAttribute, ignoreNullsExpr), aggExpr)
-        case _ => aggExpr
-      }
-    } flatMap { aggExpr =>
-      aggExpr match {
-        // We have to separate average into sum and count
-        // and for outside expression such as average(x) + 1,
-        // Spark has lift agg + 1 up to resultExpressions
-        // We need to modify the reference there as well to forge
-        // Divide(sum/count) + 1
-        case aggExpr @ AggregateExpression(Average(ref), _, _, _) =>
-          // Need a type promotion
-          val sumToPush = newAggregate(Sum(ref), aggExpr)
-          val countToPush = newAggregate(Count(ref), aggExpr)
-
-          // Need a new expression id since they are not simply rewrite as above
-          val sumFinal = newAggregateWithId(Sum(toAlias(sumToPush).toAttribute), aggExpr)
-          val countFinal = newAggregateWithId(Sum(toAlias(countToPush).toAttribute), aggExpr)
-
-          avgPushdownRewriteMap(aggExpr.resultId) = List(sumToPush, countToPush)
-          avgFinalRewriteMap(aggExpr.resultId) = List(sumFinal, countFinal)
-          List(sumFinal, countFinal)
-        case _ => aggExpr :: Nil
+        case e: Max     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: Min     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: Sum     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: First   => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case _: Count   => aggExpr.copy(aggregateFunction = Sum(partialResultRef))
+        case _: Average => throw new IllegalStateException("All AVGs should have been rewritten.")
+        case _          => aggExpr
       }
     }
 
-    val pushdownAggregates = aggregateExpressions.flatMap { aggExpr =>
-      avgPushdownRewriteMap
-        .getOrElse(aggExpr.resultId, List(aggExpr))
-    }.distinct
```

Comment:
Yes, I think so.

Reply:
This seems a problem of mine...You can leave this to me. :rofl:

Reply:
Updated.

Reply:
@liancheng Sorry to notify you that after some tests, `canonicalized aggregateExpressions` may not be able to be pushed down directly, since some expressions like `count(col_1)` will canonicalized to `count(none)` and throw unexpected exception elsewhere.

I suggest restore original non-canonicalized logic and leave this for me to fix in further PRs.

Thanks a lot!

Reply:
Thanks for the information. Reverted this change.